// bible-search/dist/resources/book.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; desc = parent = getter = undefined; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _util = require('util');

var _util2 = _interopRequireDefault(_util);

var _base = require('./base');

var _base2 = _interopRequireDefault(_base);

var inherits = _util2['default']._extend;

/**
 * Bible  Book class.
 * @class
 */

var BibleApiBook = (function (_BibleApiResource) {
  _inherits(BibleApiBook, _BibleApiResource);

  /**
   * Creates a new Bible Api Book instance.
   * @constructs BibleApiBook.
   * @param {string} apiKey - The Bible Search API Key.
   * @param {string} baseUrl - The base url (Default: https://pt-br.bibles.org/v2/).
   */

  function BibleApiBook(apiKey, baseUrl) {
    _classCallCheck(this, BibleApiBook);

    _get(Object.getPrototypeOf(BibleApiBook.prototype), 'constructor', this).call(this, apiKey, baseUrl);
  }

  /**
   * Get bible books
   * @method
   * @param {Object} args - The url arguments.
   * @param {string} args.baseUrl - The base url (Default: https://pt-br.bibles.org/v2/).
   * @param {string} args.version - Bible version (Default: por-NTLH).
   * @returns {Array} books
   */

  _createClass(BibleApiBook, [{
    key: 'getBooks',
    value: function getBooks(args) {
      args = args || {};

      var defaults = {
        baseUrl: 'https://pt-br.bibles.org/v2/',
        version: 'por-NTLH'
      };

      args = inherits(args, defaults);

      this.validate(args, ['version']);

      var url = this.getUrl({
        baseUrl: args.baseUrl,
        resource: _util2['default'].format('versions/%s/books', args.version)
      });

      return this.getJson.bind(this, url)();
    }

    /**
     * Get bible book chapters
     * @method
     * @param {Object} args - The url arguments.
     * @param {string} args.baseUrl - The base url (Default: https://pt-br.bibles.org/v2/).
     * @param {string} args.book - Book title.
     * @param {string} args.version - Bible version (Default: por-NTLH).
     * @returns {Array} chapters
     */
  }, {
    key: 'getChapters',
    value: function getChapters(args) {
      args = args || {};

      var defaults = {
        baseUrl: 'https://pt-br.bibles.org/v2/',
        version: 'por-NTLH'
      };

      args = inherits(args, defaults);

      this.validate(args, ['version', 'book']);

      var url = this.getUrl({
        baseUrl: args.baseUrl,
        resource: _util2['default'].format('books/%s:%s/chapters', args.version, args.book)
      });

      return this.getJson.bind(this, url)();
    }

    /**
     * Get bible book verses
     * @method
     * @param {Object} args - The url arguments.
     * @param {string} args.baseUrl - The base url (Default: https://pt-br.bibles.org/v2/).
     * @param {string} args.book - Book title.
     * @param {string} args.chapter - Book chapter.
     * @param {string} args.version - Bible version (Default: por-NTLH).
     * @returns {Array} verses
     */
  }, {
    key: 'getVerses',
    value: function getVerses(args) {
      args = args || {};

      var defaults = {
        baseUrl: 'https://pt-br.bibles.org/v2/',
        version: 'por-NTLH'
      };

      args = inherits(args, defaults);

      this.validate(args, ['version', 'book', 'chapter']);

      var url = this.getUrl({
        baseUrl: args.baseUrl,
        resource: _util2['default'].format('chapters/%s:%s.%s/verses', args.version, args.book, args.chapter)
      });

      return this.getJson.bind(this, url)();
    }
  }]);

  return BibleApiBook;
})(_base2['default']);

exports['default'] = BibleApiBook;
module.exports = exports['default'];
